src/libraries/System.Private.CoreLib/src/System/Threading/NamedMutex.Unix.cs

```diff
@@ -0,0 +1,655 @@
+        // independently by the processes involved. See https://github.com/dotnet/runtime/issues/10519.
+        private static bool UsePThreadMutexes => !OperatingSystem.IsApplePlatform() && !OperatingSystem.IsFreeBSD();
+
+        private readonly SharedMemoryProcessDataHeader<NamedMutexProcessDataBase> _processDataHeader = header;
+        protected nuint _lockCount;
+        private Thread? _lockOwnerThread;
+        private NamedMutexProcessDataBase? _nextInThreadOwnedNamedMutexList;
+
+        public SharedMemoryId Id => _processDataHeader._id;
+
+        public abstract bool IsLockOwnedByCurrentThread { get; }
+        public bool IsLockOwnedByAnyThreadInThisProcess => _lockOwnerThread is not null;
+
+        protected abstract void SetLockOwnerToCurrentThread();
+
+        public MutexTryAcquireLockResult TryAcquireLock(WaitSubsystem.ThreadWaitInfo waitInfo, int timeoutMilliseconds, ref WaitSubsystem.LockHolder holder)
+        {
+            SharedMemoryManager<NamedMutexProcessDataBase>.Instance.VerifyCreationDeletionProcessLockIsLocked();
+            holder.Dispose();
+            MutexTryAcquireLockResult result = AcquireLockCore(timeoutMilliseconds);
+
+            if (result == MutexTryAcquireLockResult.AcquiredLockRecursively)
+            {
+                return MutexTryAcquireLockResult.AcquiredLock;
+            }
+
+            if (result == MutexTryAcquireLockResult.TimedOut)
+            {
+                // If the lock was not acquired, we don't have any more work to do.
+                return result;
+            }
+
+            holder = SharedMemoryManager<NamedMutexProcessDataBase>.Instance.AcquireCreationDeletionProcessLock();
+            SetLockOwnerToCurrentThread();
+            _lockCount = 1;
+            _lockOwnerThread = waitInfo.Thread;
+            // Add the ref count for the thread's wait info.
+            _processDataHeader.IncrementRefCount();
+            AddOwnedNamedMutex(waitInfo.Thread, this);
+
+            if (IsAbandoned)
+            {
+                IsAbandoned = false;
+                result = MutexTryAcquireLockResult.AcquiredLockButMutexWasAbandoned;
+            }
+
+            return result;
+        }
+
+        private static void AddOwnedNamedMutex(Thread thread, NamedMutexProcessDataBase namedMutexProcessDataBase)
+        {
+            SharedMemoryManager<NamedMutexProcessDataBase>.Instance.VerifyCreationDeletionProcessLockIsLocked();
+            namedMutexProcessDataBase._nextInThreadOwnedNamedMutexList = thread.WaitInfo.LockedNamedMutexesHead;
+            thread.WaitInfo.LockedNamedMutexesHead = namedMutexProcessDataBase;
+        }
+
+        public void ReleaseLock()
+        {
+            if (!IsLockOwnedByCurrentThread)
+            {
+                throw new InvalidOperationException("Cannot release a lock that is not owned by the current thread.");
+            }
+
+            --_lockCount;
```

Comment:
Does this need to be interlocked?

Reply:
Imagine the scenario where a thread calls `ReleaseLock()` and another thread calls `Abandon()` at the same time. This could result in `_lockCount` being `-1`. Are there thread safety issues here?

Reply:
Only the owning thread can call `Abandon` or `ReleaseLock`. Any other thread calling either of those methods is invalid. I can move the assert from `ReleaseLockCore` that validates this to both callers.

Reply:
Perfect. An assert or check in `Abandon` would be appreciated. There is an explicit check in `ReleaseLock`, so I'd assume an actual check should be done there too or else both can be assert.